### PR TITLE
[Cache] Support decorated Dbal drivers in PdoAdapter

### PR DESCRIPTION
**Pull request by Jeroeny**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Explanation in this PR 
| License       | MIT

Doctrine v3 supports middleware for Drivers. Upon creating the Connection, `middleware->wrap(Driver): Driver` [is called](https://github.com/doctrine/dbal/blob/3.1.x/src/DriverManager.php#L210), in which the middleware can wrap/decorate the Driver class. So that it can perform tracing [for example](https://github.com/getsentry/sentry-symfony/blob/master/src/Tracing/Doctrine/DBAL/TracingDriverMiddleware.php#L37). 

When this happens, the Driver class inside the Connection is no longer one of Doctrine's well known Driver classes. The `PdoAdapter ` uses this class to determine the database platform. Which breaks once the Driver is decorated and no longer one of the classes [listed](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Cache/Adapter/PdoAdapter.php#L452) in the `PdoAdapter`. 

Since Dbal exposes this middleware as a feature, I think it would be nice for the `PdoAdapter` to support this.

To solve this, the `getDatabasePlatform` can be used. This returns a `Doctrine\DBAL\Platforms\AbstractPlatform` which defines the abstract method `getName`. This returns a value very similar to the list in the `PdoAdapter`. The names don't match exactly, so therefor a small mapping is done to get right the name used in the adapter. As far as a I know, there'd be no other implications with this change. 

Related: https://github.com/getsentry/sentry-symfony/issues/530

### CODE REVIEW

**Comment by derrabus** [comment]
Can you provide a test for this change, please? I think we should target 4.4 and file this as a bugfix.

**Comment by Jeroeny** [comment]
> Can you provide a test for this change, please? I think we should target 4.4 and file this as a bugfix.

Test added and rebased on 4.4.

File: src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php
Comment (Jeroeny) [inline src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php]: This is for `doctrine/dbal v2`, which doesn't have the middleware. Not sure if the test should cover this or something like a `SkippedTestSuiteError` should be thrown.

Comment (stof) [reply src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php]: I'd rather call `$this->markTestSkipped` as the feature does not exist properly for v2

File: src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php
Comment (derrabus) [inline src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php]: Let‘s move the class into a dedicated file in the `Tests\Fixtures` namespace, so we don‘t need a conditional declaration.

File: src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php
Comment (derrabus) [inline src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php]: Please use an import for `Middleware`.

File: src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php
Comment (derrabus) [reply src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php]: I agree. If the test is irrelevant for DBAL 2, just skip it if we‘re not testing against DBAL 3.

Comment (Jean85) [reply src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php]: > I agree. If the test is irrelevant for DBAL 2, just skip it if we‘re not testing against DBAL 3.

I don't think that this fix is irrelevant for DBAL 2, because you could still decorate/wrap the driver manually anyway, see #42022 and the originating issue https://github.com/getsentry/sentry-symfony/issues/530.

In general, I think we should change the `switch` to use `$this->conn->getDatabasePlatform()->getName()` as a reliable source of information.

**Review by Jean85 — CHANGES_REQUESTED** [review]
Thanks for this PR! This should fix #42022

I don't understand why you didn't use `$this->conn->getDatabasePlatform()->getName()` as you suggested yourself though...

**Comment by Jeroeny** [comment]
> I don't understand why you didn't use `$this->conn->getDatabasePlatform()->getName()` as you suggested yourself though...

I did though? See [`PdoTrait:458`](https://github.com/symfony/symfony/pull/42011/files#diff-a508f4ce64bb6d2ab3c496c220fb866aade4e777a464817576f4be0bed306b4dR458)  

Edit: I see there's a slight difference (`$driver->getDatabasePlatform()->getName()`), but I'm not sure how important that difference is?

File: src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php
Comment (Jeroeny) [reply src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php]: > because you could still decorate/wrap the driver manually anyway

But not as official dbal v2 functionality right? You can pass a `driverClass` to the configuration, which is then constructed without arguments, making it less usable to function as decorator. Or it'd be via Reflection changing the private property `_driver` to public at runtime. Which is in my opinion already such an edge case, that Symfony shouldn't have to test against that.

File: src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php
Comment (Jeroeny) [reply src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php]: It seems `php .github/patch-types.php` is not happy with this class existing as fixture: https://github.com/symfony/symfony/pull/42011/checks?check_run_id=3016796261 (Or is something else wrong?)

Comment (derrabus) [reply src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php]: patch-types has a list of files to exclude. Please add your fixture there.

Comment (Jeroeny) [reply src/Symfony/Component/Cache/Tests/Adapter/PdoDbalAdapterTest.php]: Thanks

**Comment by Jean85** [comment]
> > I don't understand why you didn't use `$this->conn->getDatabasePlatform()->getName()` as you suggested yourself though...
> 
> I did though? See [`PdoTrait:458`](https://github.com/symfony/symfony/pull/42011/files#diff-a508f4ce64bb6d2ab3c496c220fb866aade4e777a464817576f4be0bed306b4dR458)
> 
> Edit: I see there's a slight difference (`$driver->getDatabasePlatform()->getName()`), but I'm not sure how important that difference is?

My suggestion is to change [the whole `switch`](https://github.com/symfony/symfony/pull/42011/files#diff-a508f4ce64bb6d2ab3c496c220fb866aade4e777a464817576f4be0bed306b4dR427) to drive the choice using `$driver->getDatabasePlatform()->getName()`, because the `instaceof` will be broken by wrappers.

**Comment by Jeroeny** [comment]
> My suggestion is to change [the whole `switch`](https://github.com/symfony/symfony/pull/42011/files#diff-a508f4ce64bb6d2ab3c496c220fb866aade4e777a464817576f4be0bed306b4dR427) to drive the choice using `$driver->getDatabasePlatform()->getName()`, because the `instaceof` will be broken by wrappers.

That's good feedback 👍 . Though I'd like to get some feedback on that from a Symfony maintainer as well, if possible. They're the ones to approve the PR after all.

**Comment by Jean85** [comment]
> > My suggestion is to change [the whole `switch`](https://github.com/symfony/symfony/pull/42011/files#diff-a508f4ce64bb6d2ab3c496c220fb866aade4e777a464817576f4be0bed306b4dR427) to drive the choice using `$driver->getDatabasePlatform()->getName()`, because the `instaceof` will be broken by wrappers.
> 
> That's good feedback +1 . Though I'd like to get some feedback on that from a Symfony maintainer as well, if possible. They're the ones to approve the PR after all.

@nicolas-grekas already weighed in in https://github.com/symfony/symfony/issues/42022#issuecomment-876089615, but I'm not sure if he got the full implication of my suggestion.

File: .github/patch-types.php
Comment (nicolas-grekas) [inline .github/patch-types.php:28]: why that? can't you add the return types to the file instead?

**Comment by nicolas-grekas** [comment]
> change the whole switch

That works for me if the resulting code remains compatible with the range of dbal versions we support currently.

File: .github/patch-types.php
Comment (Jeroeny) [reply .github/patch-types.php:28]: The `patch-types` runs with DBAL v2 installed, whereas this `DriverWrapper` class implements the interface of v3. Tested it again with return types to be sure: https://github.com/symfony/symfony/pull/42011/checks?check_run_id=3054146971

**Review by derrabus — CHANGES_REQUESTED** [review]
In patch-types.php, we maintain a list of files that should be ignored by that script. Please add your new `DriverWrapper` fixture there to make the tests pass.

File: .github/patch-types.php
Comment (Jeroeny) [reply .github/patch-types.php:28]: Since I couldn't make that work (it would mean making the `DriverWrapper` compatible with both v2 and v3), and the `DriverWrapper` is only loaded for the dbal v3 tests, I've re-added this exclusion. I hope that's okay.

**Review by derrabus — APPROVED** [review]

**Comment by derrabus** [comment]
LGTM, but I'd turn the `@return` annotations of the `DriverWrapper` fixture into proper return types. This way, it does not matter that much if we exclude it from the patch-types script.

**Review by Tobion — APPROVED** [review]

**Comment by Tobion** [comment]
Thank you @Jeroeny.

**Pull request merged**
